### ISSUE TITLE
ECIP-?-smart contract based system for ECIP management

### ISSUE THREAD

**Issue by realcodywburns** [ethereumproject/EIPs#42]
### A smart contract based system for ECIP management

    ECIP: TBD
    Title: A smart contract based system for ECIP management
    Status: Draft
    Type: Meta
    Author: Cody.W.Burns@ethereumclassic.org
    Created: 2017-03-26

### Abstract
ECIP stands for Ethereum Improvement Proposal. An ECIP is a design document providing information to the Ethereum classic community, or describing a new feature for Ethereum or its processes or environment. The ECIP should provide a concise technical specification of the feature and a rationale for the feature.  The ECIP author is responsible for building consensus within the community and documenting dissenting opinions. This system is currently managed using a github repository which does not allow for voting. By using a smart contract based system for tracking and managing ECIPs it is hope a broader participation from the community can be achieved. 

### Motivation
To overarching motivation for this proposal is to provide a medium for all future ECIP's a succinct environment allowing for financial transparency, a more inclusive decision making process, a disintermediation of proposals from the proposers and better funding channels to avoid centralization

#### ECIP Types
There are three kinds of ECIP:

* A Standard Track ECIP describes any change that affects most or all Ethereum implementations, such as a change to the [https://github.com/ethereum/yellowpaper Yellow Paper], the network protocol, a change in block or transaction validity rules, or any change or addition that affects the interoperability of applications using Ethereum. Furthermore Standard ECIPs can be broken down into the following categories.
** Consensus - Those ECIP's which effect the system of block creation
** Networking - Currently Networking discussion tracks in the [https://github.com/ethereum/devp2p devp2p repository].

* An Informational ECIP describes a Ethereum Classic design issue, or provides general guidelines or information to the ETC community, but does not propose a new feature. Informational ECIPs do not necessarily represent Ethereum community consensus or a recommendation, so users and implementers are free to ignore Informational ECIPs or follow their advice.
* A Meta ECIP describes a process surrounding Ethereum or proposes a change to (or an event in) a process. Process ECIPs are like Standards Track ECIPs but apply to areas other than the Ethereum protocol itself. They may propose an implementation, but not to Ethereum's codebase; they often require community consensus; unlike Informational ECIPs, they are more than recommendations, and users are typically not free to ignore them. Examples include procedures, guidelines, changes to the decision-making process, and changes to the tools or environment used in Ethereum development. Any meta-ECIP is also considered a Process ECIP.

#### Definitions

Registry: A publicly accessible smart contract that allows individual ECIPs the ability to register. It should be robust enough to provide:
 * Basic administration of registered contracts
 * Allow for status tracking registered contracts
 * Allow for typing to registered contracts.
 * It should provide a basic system framework for voting 
 * Allow for voter registration
 * Funding management as required.(RFC)
 * Maintain a publicly accessible log of administrator activity

ECIP contract: Smart contract that contains meta-data relating to the ECIP. At a minimum it should:
  * contain meta-data of the proposed ECIP
  * Allow for commenting on contract

Front-end system: Any system external to the ECIP contracts which is able to submit web3(or similar) calls to the ETC EVM to retrieve information from contracts and present the data in a user friendly interface
ECIP collaborators – trusted community members assigned as administrators to the registry. 

Legacy ECIP Process: 
#### Proposed ECIP Work Flow

The ECIP process begins with a new idea for Ethereum Classic. It is highly recommended that a single ECIP contain a single key proposal or new idea. Small enhancements or patches that don't affect consensus often don't need an ECIP and can be injected into the ETC development workflow with a patch submission to the corresponding ETC issue tracker. The more focused the ECIP, the more successful it tends to be. The ECIP editor reserves the right to reject ECIP proposals if they appear too unfocused or too broad. If in doubt, split your ECIP into several well-focused ones.

Each ECIP must have an Owner -- someone who writes the ECIP using the style and format described below, submits the smart contract to allow for discussion, and attempts to build community consensus around the idea. The ECIP champion (a.k.a. Owner) should first attempt to ascertain whether the idea is ECIP-able.

Vetting an idea publicly before going as far as writing an ECIP is meant to save the potential author time. Asking the ETC community first if an idea is original helps prevent too much time being spent on something that is guaranteed to be rejected based on prior discussions (searching the Internet does not always do the trick). It also helps to make sure the idea is applicable to the entire community and not just the author. Just because an idea sounds good to the author does not mean it will work for most people in most areas where ETC is used.

Once the owner has asked the ETC community as to whether an idea has any chance of acceptance, a draft ECIP should be registered on the ETC chain using a standard ECIP contract and the ECIP registry. This gives the author a chance to flesh out the draft ECIP to make properly formatted, of high quality, and to address initial concerns about the proposal.

Once submitted to the registry contract the ECIP will be reviewed by the ECIP collaborators for fitness, grammar, and purpose the ECIP registry will assign the ECIP a number, label it as Standards Track, Informational, or Process, give it status "Draft", and add it to the git repository. The ECIP editor will not unreasonably deny an ECIP. Reasons for denying ECIP status include duplication of effort, being technically unsound, not providing proper motivation or addressing backwards compatibility, or not in keeping with the ETC philosophy.

The ECIP author may update the Draft as necessary in their smart contract. The ECIP itself should not be stored on chain.  
Standards Track ECIPs consist of three parts, a design document, implementation and finally if warranted an update to the [https://github.com/ethereum/yellowpaper formal specification]. The ECIP should be reviewed and accepted before an implementation is begun, unless an implementation will aid people in studying the ECIP. Standards Track ECIPs must be implemented in at least two viable Ethereum clients before it can be considered Final.

ECIP authors are responsible for collecting community feedback on an ECIP before submitting it for review. However, wherever possible, long open-ended discussions should be avoided. Strategies to keep the discussions efficient include: having the ECIP author accept private comments in the early design phases, setting up a wiki page or git repository, etc. ECIP authors should use their discretion here. Whenever possible, set an end date or block number for final comment. 

For an ECIP to be accepted it must meet certain minimum criteria. It must be a clear and complete description of the proposed enhancement. The enhancement must represent a net improvement. The proposed implementation, if applicable, must be solid and must not complicate the protocol unduly.

Once a ECIP has been accepted, the implementations must be completed. When the implementation is complete in at least two viable clients and accepted by the community, the status will be changed to "Final". An update to the [https://github.com/ethereum/yellowpaper formal specification] should accompany the "Final" status change.

An ECIP can also be assigned status "Deferred". The ECIP author or editor can assign the ECIP this status when no progress is being made on the ECIP. Once a ECIP is deferred, the ECIP editor can re-assign it to draft status.

A ECIP can also be "Rejected". Perhaps after all is said and done it was not a good idea. It is still important to have a record of this fact.

ECIPs can also be superseded by a different ECIP, rendering the original obsolete. This is intended for Informational ECIPs, where version 2 of an API can replace version 1.
Some Informational and Process ECIPs may also have a status of "Active" if they are never meant to be completed. E.g. ECIP 1 (this ECIP).

#### What belongs in an ECIP contract?
Each ECIP should have the following parts:

* Title: <short title of ECIP>

* Author: <list of authors' real names and optionally, email address>

* Description: < short description of the ECIP >

* External Url: <location of smart contract>

* Comment array:  < string array for on contract comment tracking>

#### Transferring ECIP Ownership
It occasionally becomes necessary to transfer ownership of ECIPs to a new champion. In general, we'd like to retain the original author as a co-author of the transferred ECIP, but that's really up to the original author. A good reason to transfer ownership is because the original author no longer has the time or interest in updating it or following through with the ECIP process, or has fallen off the face of the 'net (i.e. is unreachable or not responding to email). A bad reason to transfer ownership is because you don't agree with the direction of the ECIP. We try to build consensus around a ECIP, but if that's not possible, you can always submit a competing ECIP.

If you are interested in assuming ownership of a ECIP, send a message asking to take over, addressed to both the original author and the ECIP editor. If the original author doesn't respond to email in a timely manner, the ECIP editor will make a unilateral decision (it's not like such decisions can't be reversed :).

#### ECIP Editors
The current ECIP editors are: 
 * TBD
#### ECIP Editor Responsibilities & Workflow
For each new ECIP that comes in, an editor does the following:

* Read the ECIP to check if it is ready: sound and complete. The ideas must make technical sense, even if they don't seem likely to be accepted.
* The title should accurately describe the content.
* Edit the ECIP for language (spelling, grammar, sentence structure, etc.), markup (for reST EWIPs), code style

If the ECIP isn't ready, the editor will send it back to the author for revision, with specific instructions.

Once the ECIP is ready for the repository, the ECIP editor will:

* Assign a ECIP number (almost always just the next available number)

* Assign the appropriate fields in the registry for type, tract, and status

* Send email back to the ECIP author with next step.

Many ECIPs are written and maintained by developers with write access to the ETC codebase. The ECIP editors monitor ECIP changes, and correct any structure, grammar, spelling, or markup mistakes they see.

The editors don't pass judgment on ECIPs. They merely do the administrative & editorial part. Except for times like this, there's relatively low volume.

### Voting
…

### Implementation
A reference ECIP smart contract  can be found at [[ 0xd9605C12abc5Bfb83bbab80d286821D110dA2dD3 ]]
JSON-ABI:
[{"constant":false,"inputs":[{"name":"_title","type":"string"},{"name":"_author","type":"string"},{"name":"_description","type":"string"},{"name":"_extUrl","type":"string"},{"name":"_type","type":"string"}],"name":"startecip","outputs":[],"payable":true,"type":"function"},{"constant":true,"inputs":[],"name":"created","outputs":[{"name":"","type":"uint256"}],"payable":false,"type":"function"},{"constant":true,"inputs":[],"name":"extUrl","outputs":[{"name":"","type":"string"}],"payable":false,"type":"function"},{"constant":false,"inputs":[],"name":"kill","outputs":[],"payable":false,"type":"function"},{"constant":true,"inputs":[],"name":"title","outputs":[{"name":"","type":"string"}],"payable":false,"type":"function"},{"constant":false,"inputs":[],"name":"bailout","outputs":[],"payable":false,"type":"function"},{"constant":true,"inputs":[],"name":"description","outputs":[{"name":"","type":"string"}],"payable":false,"type":"function"},{"constant":false,"inputs":[{"name":"_comment","type":"string"}],"name":"comment","outputs":[],"payable":false,"type":"function"},{"constant":true,"inputs":[],"name":"author","outputs":[{"name":"","type":"string"}],"payable":false,"type":"function"},{"anonymous":false,"inputs":[{"indexed":false,"name":"comment","type":"string"}],"name":"newComment","type":"event"}]

The ECIP master registry contract is located at [[ ]]
### History
This document was derived heavily from [https://github.com/bitcoin/bips Bitcoin's BIP-0001] written by Amir Taaki which in turn was derived from [https://www.python.org/dev/peps/ Python's PEP-0001]. In many places text was simply copied and modified. Although the PEP-0001 text was written by Barry Warsaw, Jeremy Hylton, and David Goodger, they are not responsible for its use in the Ethereum Classic Improvement Process, and should not be bothered with technical questions specific to ETC or the ECIP. Please direct all comments to the ECIP editors.

**Comment by whysoserious** [comment]
Hi Cody, I am Jan from Grothendieck team. I am working on a very similar thing as you: [New-ECIP-Proposal](https://github.com/input-output-hk/ecip-quorum/wiki/New-ECIP-Proposal). It's going to change very soon as this is work in progress. However [TL;DR](https://github.com/input-output-hk/ecip-quorum/wiki/New-ECIP-Proposal#new-ecip-work-flow) section should give you a big picture of what we are working on.
Hopefully we'll be able to merge both docs into one document very soon.

For now I have only one issue to comment. 

In my opinion a draft of an ECIP and comments should be stored off-chain. Blockchain should be used only for voting and managing status of ECIPs. I see no benefit of paying gas for adding a comment. Thus, initial list of fields of an ECIP proposal would look as follows:
* **Title**:
* **Author**: <list of authors' real names and optionally, email address>
* **Description**: < short description of the ECIP >
* **URL to the full proposal description**:
* **Hash of the full proposal description**: `curl -s URL-TO-THE-FULL-PROPSAL-DESC | md5`
* ~~Comment array: < string array for on contract comment tracking>~~
* *more fields to follow - eg. start/end block when voting is allowed, status*

**Comment by realcodywburns** [comment]
I don't disagree. It would greatly reduce the complexity and chain weight to keep general comments offchain

**Comment by laddhoffman** [comment]
Going to jump in here and share my thoughts.

GitHub issues provide a convenient way to tag the status of issues, and it is easy for anyone to comment. This is good as it lends itself to a transparent process, at least informally.

GitHub repositories provide a convenient way for an author to publish a draft proposal and for others to submit pull requests for modifications to the proposal.

Editing, rather than a role of authority, can be considered a service that people perform. It can be instrumented through submitting pull requests containing enhancements, such as corrections to grammar. Objections to content can be added as comments to the associated GitHub issue, visible to all community members.
An author seeking community adoption of a proposal will then bear the responsibility for incorporating the feedback provided by members of the community for the sake of improving the proposal.

It sounds like ETC smart contracts can provide a good mechanism for voting. I'm not familiar with the details yet. But unless we are talking about some kind of DAO, I gather that the purpose of the on-blockchain voting is to provide a form of documentation to support followup actions? I.e. in the case of a technical ECIP, an eventual pull request submitted to the ethereum classic codebase would be accompanied by a reference to the vote.

Jan proposes to include the URL to a proposal and a hash of its contents in the on-blockchain ECIP record. However, in the event that the contents are in fact modified, this does not provide a mechanism for recovering the original contents. To mitigate this, I propose that the ECIP record refer to a specific git repository and commit id.

The voting mechanism may always be something of an approximation to consensus, because at best it can only capture a snapshot. i.e. agreement with a particular version of a proposal. In the end, each person involved makes decisions based on their interests and their understanding of the information available. So assuming ethical actors the system should optimize for efficiently getting an accurate picture of what has occured. And assuming malicious actors, the system should provide transparency so that records can be verified by anyone.

Example story:

The community discusses A, B, and C. Sam proposes D.0. The community provides input, and Sam incorporates it along with his own work to produce D.1. The community votes to approve D.1 with 60% participation and 70% of those who participate are in agreement. Sam and Kathy then implement software changes, but update the proposal to D.2 in light of new understanding. Now the ethereum classic code maintainers are reviewing a pull request for the implementation of D.2 and they have the choice to accept it, or to ask that a new vote be taken?

**Comment by realcodywburns** [comment]
Closing, this repo is depreciated . To be reopened on http://github.com/ethereumclassic/ECIPs
